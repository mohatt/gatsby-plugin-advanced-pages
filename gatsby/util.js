const path = require('path')
const _ = require("lodash")
const debug = require("debug")("gatsby-plugin-advanced-pages")

// Initializes options
const initializeOptions = _.once(({ store, reporter, defaultOptions, userOptions }) => {
  const { program } = store.getState()
  const o = _.merge(defaultOptions, userOptions)

  // Ensure basePath is absolute
  o.basePath = path.join('/', o.basePath)

  // Ensure type names are Capitalized
  o.typeNames = _.mapValues(o.typeNames, _.upperFirst)

  // Generate mdxFrontmatter type name based on page type
  o.typeNames.mdxFrontmatter = 'Mdx' + o.typeNames.page + 'Frontmatter'

  // Convert directory paths to absolute ones
  o.directories = _.mapValues(o.directories, dir => (
    path.join(program.directory, dir)
  ))

  // Make sure we have a valid positive number
  o.pagination.limit = parseInt(o.pagination.limit)
  if(o.pagination.limit <= 0 || isNaN(o.pagination.limit)){
    throw new TypeError(`Expected a positive number for 'pagination.limit' option`)
  }

  // Debug final options object
  debug("Options", o)

  return o
})

// Gets the initialized options object
const getOptions = () => {
  return initializeOptions()
}

// Gets the value of a single option
const getOption = (optionName) => {
  return _.get(getOptions(), optionName)
}

// Checks if a path is child of another one
const isPathChildof = (child, parent) => {
  const relative = path.relative(parent, child)
  return relative && !relative.startsWith('..') && !path.isAbsolute(relative)
}

module.exports = {
  debug,
  initializeOptions,
  getOptions,
  getOption,
  isPathChildof
}
