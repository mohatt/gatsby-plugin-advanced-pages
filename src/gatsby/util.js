import fs from 'fs'
import path from 'path'
import _ from 'lodash'
import debug from 'debug'

// Initializes options
// The function only runs when called with args to initialize and cache
// options object. Subsequent calls without args return the cached value
export function initializeOptions (args) {
  if (arguments.length === 0) {
    if (typeof initializeOptions.options === 'undefined') {
      return reportError('Cant fetch options because they are not initialized')
    }

    return initializeOptions.options
  }

  // Merge user-defined options with defaults
  const options = _.merge(args.defaultOptions, args.pluginOptions)

  // set the root path of the the project
  options._root = args.store.getState().program.directory

  // Ensure basePath is absolute
  options.basePath = path.join('/', options.basePath)

  // Cache the options object now so that subsequent
  // calls dont throw an error
  initializeOptions.options = options

  // Verify directory locations
  options.directories = _.mapValues(options.directories, dir => {
    return lookupPath(dir)
  })

  // Verify default template
  if(options.template) {
    options.template = lookupPath(options.template, options.directories.templates)
  }

  // Debug final options object
  debug('gatsby-plugin-advanced-pages')('Options', options)

  initializeOptions.options = options
}

// Initializes reporter
// The function only runs when called with args to initialize and cache
// reporter object. Subsequent calls without args return the cached value
export function initializeReporter (reporter) {
  if (arguments.length === 0) {
    if (typeof initializeReporter.reporter === 'undefined') {
      throw new Error('Cant fetch reporter because it is not initialized')
    }

    return initializeReporter.reporter
  }

  const errorMap = {
    [10000]: {
      text: context => context.message,
      level: `ERROR`,
      type: `PLUGIN`,
    },
  }

  if (reporter.setErrorMap) {
    reporter.setErrorMap(errorMap)
  }

  initializeReporter.reporter = reporter
}

// Gets the initialized eporter object
export function getReporter () {
  return initializeReporter()
}

// Gets the initialized eporter object
export function reportError (message, e = null) {
  const reporter = getReporter()
  const prefix = `"gatsby-plugin-advanced-pages" threw an error while running`

  if(!e) {
    reporter.panic({
      id: `10000`,
      context: {
        message: prefix
      },
      error: new Error(message)
    })
    return
  }

  reporter.panic({
    id: `10000`,
    context: {
      message: `${prefix}:\n ${message}`
    },
    error: e
  })
}

// Gets the initialized options object
export function getOptions () {
  return initializeOptions()
}

// Gets the value of a single option
export function getOption (optionName) {
  return _.get(getOptions(), optionName)
}

// Checks if a location exists under parent
// if not checks if it exists under project root
// if not check if its an absolute path
// throws an error if cant find any
export function lookupPath(location, parent = null){
  if(!location || typeof location !== 'string') {
    return reportError(
      `ensurePath() expected a non-empty string but got ${typeof location}("${location}")`
    )
  }

  let search = []
  if(parent) {
    const localPath = path.join(parent, location)
    search.push(localPath)
    if(fs.existsSync(localPath)) {
      return localPath
    }
  }

  const rootPath = path.join(getOption('_root'), location)
  search.push(rootPath)
  if(fs.existsSync(rootPath)) {
    return rootPath
  }

  search.push(location)
  if(fs.existsSync(location)) {
    return location
  }

  reportError(
    `A path with value "${location}" could not be found at ` +
    `any of the following locations:\n - "${search.join(`\n - "`)}"`
  )
}
