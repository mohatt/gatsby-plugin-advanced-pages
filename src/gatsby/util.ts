import fs from 'fs'
import path from 'path'
import _ from 'lodash'
import debug from 'debug'
import type { NodePluginArgs, Reporter } from 'gatsby'
import type { PluginOptions } from './types'

export interface ResolvedPluginOptions extends PluginOptions {
  _root: string;
}

/**
 * Represents the internal state of the plugin.
 */
const internalState: {
  options?: ResolvedPluginOptions
  reporter?: Reporter
} = {}

// Initializes options
// The function only runs when called with args to initialize and cache
// options object. Subsequent calls without args return the cached value
export const initializeOptions = (args?: Pick<NodePluginArgs, 'store'> & { pluginOptions: Partial<PluginOptions>, defaultOptions: PluginOptions }): ResolvedPluginOptions => {
  if (args === undefined) {
    if (!('options' in internalState)) {
      reportError('Cant fetch options because they are not initialized')
      return null
    }

    return internalState.options
  }

  // Merge user-defined options with defaults
  const options: ResolvedPluginOptions = _.merge(
    // set the root path of the project
    { _root: args.store.getState().program.directory },
    args.defaultOptions,
    args.pluginOptions,
  )
  // Ensure basePath is absolute
  options.basePath = path.join('/', options.basePath)
  // Ensure type names are Title case
  options.typeNames = _.mapValues(options.typeNames, _.upperFirst)
  // Convert directory paths to absolute ones
  options.directories = _.mapValues(options.directories, dir => (
    path.join(options._root, dir)
  ))

  // Cache the options object now so that subsequent
  // calls to lookupPath() dont throw an error
  internalState.options = options

  // Verify default template
  if (options.template) {
    options.template = lookupPath(options.template, options.directories.templates)
  }

  // Debug final options object
  debug('gatsby-plugin-advanced-pages')('Options', options)

  internalState.options = options

  return options
}

// Initializes reporter
// The function only runs when called with args to initialize and cache
// reporter object. Subsequent calls without args return the cached value
export const initializeReporter = (reporter?: Reporter) => {
  if (reporter === undefined) {
    if (!('reporter' in internalState)) {
      throw new Error('Cant fetch reporter because it is not initialized')
    }

    return internalState.reporter
  }

  if (reporter.setErrorMap) {
    reporter.setErrorMap({
      10000: {
        text: context => context.message,
        level: 'ERROR',
        type: 'PLUGIN'
      }
    })
  }

  internalState.reporter = reporter
  return reporter
}

// Gets the initialized reporter object
export const getReporter = () => {
  return initializeReporter()
}

// Prints an error message and terminates the build
export const reportError = (message: string, error?: Error): never => {
  const reporter = getReporter()
  const prefix = '"gatsby-plugin-advanced-pages" threw an error while running'

  if (!error) {
    return reporter.panic({
      id: '10000',
      context: {
        message: prefix
      },
      error: new Error(message)
    })
  }

  return reporter.panic({
    id: '10000',
    context: {
      message: `${prefix}:\n ${message}`
    },
    error
  })
}

// Prints a warning message
export const reportWarning = (message: string) => {
  const reporter = getReporter()
  const prefix = '"gatsby-plugin-advanced-pages" might not be working properly'
  reporter.warn(`${prefix}:\n ${message}`)
}

// Gets the initialized options object
export const getOptions = () => {
  return initializeOptions()
}

// Gets the value of a single option
export const getOption = <T extends keyof ResolvedPluginOptions>(optionName: T): ResolvedPluginOptions[T] => {
  return _.get(getOptions(), optionName)
}

// Checks if a location exists under parent
// if not checks if it exists under project root
// if not check if its an absolute path
// throws an error if cant find any
export const lookupPath = (location: string, parent?: string) => {
  if (!location || typeof location !== 'string') {
    reportError(
      `ensurePath() expected a non-empty string but got ${typeof location}("${location}")`
    )
    return null
  }

  const search = []
  if (parent && typeof parent === 'string') {
    const localPath = path.join(parent, location)
    search.push(localPath)
    if (fs.existsSync(localPath)) {
      return localPath
    }
  }

  const rootPath = path.join(getOption('_root'), location)
  search.push(rootPath)
  if (fs.existsSync(rootPath)) {
    return rootPath
  }

  if (path.isAbsolute(location)) {
    search.push(location)
    if (fs.existsSync(location)) {
      return location
    }
  }

  reportError(
    `A path with value "${location}" could not be found at ` +
    `any of the following locations:\n - "${search.join('\n - "')}"`
  )
  return null
}
