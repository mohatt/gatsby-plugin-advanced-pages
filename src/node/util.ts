import fs from 'fs'
import path from 'path'
import { createRequire } from 'module'
import Yaml from 'js-yaml'
import { validateOptionsSchema, Joi, PluginOptionsSchemaJoi, Schema } from 'gatsby-plugin-utils'
import _ from 'lodash'
import debug from 'debug'
import type { GatsbyNode, NodePluginArgs, Reporter } from 'gatsby'
import type { DefaultPluginOptions, PluginOptions } from './api'

/**
 * Represents the resolved internal options for the plugin.
 * It includes additional properties required for internal processing.
 */
interface ResolvedPluginOptions extends DefaultPluginOptions {
  /**
   * Absolute path to the root of the Gatsby project.
   * Used to resolve relative paths for templates, directories, etc.
   */
  _root: string
}

/**
 * Provides functionality to initialize, merge, and retrieve plugin options.
 *
 * @internal
 */
export const options = (() => {
  let ref: ResolvedPluginOptions | undefined

  /**
   * Initializes and merges plugin options with defaults.
   */
  const initialize = (args: {
    store: NodePluginArgs['store']
    pluginOptions: PluginOptions
    defaultOptions: DefaultPluginOptions
  }): ResolvedPluginOptions => {
    // Merge user-defined options with defaults
    const opts: ResolvedPluginOptions = _.merge(
      // set the root path of the project
      { _root: args.store.getState().program.directory },
      args.defaultOptions,
      args.pluginOptions,
    )
    // Normalize options
    Object.assign(opts, {
      // Ensure basePath is absolute
      basePath: path.join('/', opts.basePath),
      // Ensure type names are Title case
      typeNames: _.mapValues(opts.typeNames, _.upperFirst),
      // Convert directory paths to absolute ones
      directories: _.mapValues(opts.directories, (dir) => path.join(opts._root, dir)),
    })

    // Cache the options object now so that subsequent
    // calls to ensurePath() don't throw an error
    ref = opts

    // Verify default template exists
    if (opts.template != null) {
      opts.template = ensurePath(opts.template, opts.directories.templates)
    }

    // Debug final options object
    debug('gatsby-plugin-advanced-pages')('Options', opts)

    ref = opts

    return opts
  }

  /**
   * Retrieves a specific option value or the entire options object.
   *
   * - If called without arguments, returns the entire options object.
   * - If called with a key, retrieves the corresponding value.
   * - Throws an error if options are not initialized.
   *
   * @param optionName (Optional) The key of the option to retrieve.
   *
   * @returns The requested option value or the entire options object.
   *
   * @throws - Terminates execution if options are not initialized.
   */
  const get: {
    (): ResolvedPluginOptions
    <T extends keyof ResolvedPluginOptions>(optionName: T): ResolvedPluginOptions[T]
  } = (...args: [] | [string]) => {
    if (ref === undefined) {
      // throw new Error('Plugin options are not initialized.')
      throw new PluginError('Cant fetch options because they are not initialized')
    }

    if (args.length === 0) {
      return ref
    }

    return _.get(ref, ...args)
  }

  return { initialize, get }
})()

/**
 * Manages Gatsby's reporter instance, providing functionalities for structured error logging,
 * warnings, and instance management.
 *
 * @internal
 */
export const reporter = (() => {
  let ref: Reporter | undefined

  // Initializes reporter
  /**
   * Initializes and caches the Gatsby reporter instance.
   *
   * If `setErrorMap` is available, it registers custom error messages
   * for better debugging in Gatsby's CLI.
   */
  const initialize = (instance: Reporter): Reporter => {
    if (instance.setErrorMap != null) {
      instance.setErrorMap({
        10000: {
          text: (context) => context.message,
          category: 'USER',
          level: 'ERROR',
          type: 'PLUGIN',
        },
      })
    }

    ref = instance
    return instance
  }

  /**
   * Retrieves the cached reporter instance.

   * @throws Error if the reporter is accessed before being initialized.
   */
  const get = (): Reporter => {
    if (ref === undefined) {
      throw new Error('Cant fetch reporter instance because it is not initialized')
    }

    return ref
  }

  /**
   * Logs an error and terminates the build process.
   *
   * - Uses Gatsby's `panic()` method to log structured errors.
   * - If an `Error` object is provided, includes it in the error report.
   * - Otherwise, creates a new error instance from the given message.
   *
   * @param api The Gatsby Node API that caused the error.
   *
   * @param titleOrError The error message or PluginError object to display.
   *
   * @param error (Optional) The original error instance.
   *
   * @throws - This function never returns; it always terminates execution.
   */
  const error = (
    api: keyof GatsbyNode,
    titleOrError: string | PluginError | Error,
    error?: Error,
  ): never => {
    const instance = get()
    const prefix = `"gatsby-plugin-advanced-pages" threw an error during "${api}" hook`

    let title: string | undefined
    let mainError: Error | undefined

    if (titleOrError instanceof PluginError) {
      title = titleOrError.message
      mainError = titleOrError.originalError
    } else if (titleOrError instanceof Error) {
      mainError = titleOrError
    } else {
      title = titleOrError
    }

    if (error) {
      mainError = error
    }

    return instance.panic({
      id: '10000',
      context: {
        message: mainError && title ? `${prefix}:\n ${title}` : prefix,
      },
      error: mainError ?? (title ? new Error(title) : undefined),
    })
  }

  /**
   * Logs a warning message in the Gatsby CLI.
   *
   * @param message The warning message to display.
   */
  const warning = (message: string): void => {
    const instance = get()
    const prefix = '"gatsby-plugin-advanced-pages" might not be working properly'
    instance.warn(`${prefix}:\n ${message}`)
  }

  return { initialize, get, error, warning }
})()

/**
 * Validates the input data using Joi schema validation.
 *
 * @param schema The Joi schema (or schema factory) to validate against.
 * @param data The input data to validate.
 * @param [name] (Optional) The prop name of the data being validated. Default is `data`.
 *
 * @returns A promise that resolves to the validated and typed data.
 *
 * @throws {ValidationError} Throws an error if the input does not meet the schema requirements.
 */
export const validateSchema = async <E>(
  schema: Schema | ((Joi: PluginOptionsSchemaJoi) => Schema),
  data: unknown,
  name = 'data',
): Promise<E> => {
  const objSchema = Joi.object({ [name]: schema instanceof Function ? schema(Joi) : schema })
  const { value } = await validateOptionsSchema(objSchema, { [name]: data })
  return value[name] as E
}

/**
 * Ensures a path exists based on different search locations:
 *
 * The function checks in the following order:
 * 1. If `location` exists under the given `parent` directory, return it.
 * 2. If not found, checks under the Gatsby project root (`_root`).
 * 3. If still not found and `location` is absolute, it verifies the path exists.
 * 4. If none of the above match, logs an error and terminates execution.
 *
 * @param location The relative or absolute file path to resolve.
 * @param parent (Optional) A parent directory to check first.
 *
 * @returns The absolute path if found; otherwise, throws an error.
 *
 * @throws PluginError if the path is not found.
 */
export const ensurePath = (location: string, parent?: string): string => {
  const search = []
  if (parent != null) {
    const localPath = path.join(parent, location)
    search.push(localPath)
    if (fs.existsSync(localPath)) {
      return localPath
    }
  }

  const rootPath = path.join(options.get('_root'), location)
  search.push(rootPath)
  if (fs.existsSync(rootPath)) {
    return rootPath
  }

  if (path.isAbsolute(location)) {
    search.push(location)
    if (fs.existsSync(location)) {
      return location
    }
  }

  throw new PluginError(
    `A path with value "${location}" could not be found at ` +
      `any of the following locations:\n - "${search.join('\n - "')}"`,
  )
}

/**
 * Retrieves the default export of a module.
 */
export const getDefaultExport = async <T = unknown>(modulePath: string): Promise<T> => {
  const noop = {}
  let defaultExport: any = noop

  // Avoid require in tests since Vitest doesn't support mocking require calls
  if (process.env.VITEST !== 'true') {
    const nodeRequire = require ?? createRequire(import.meta.url)
    try {
      defaultExport = nodeRequire(modulePath)
      if (defaultExport?.__esModule === true && 'default' in defaultExport) {
        defaultExport = defaultExport.default
      }
    } catch (error) {
      if (error.code !== 'ERR_REQUIRE_ESM') {
        throw new PluginError(`Unable to import module "${modulePath}"`, error)
      }
    }
  }

  if (defaultExport === noop) {
    try {
      const esmExports = await import(modulePath)
      defaultExport = esmExports.default
    } catch (error) {
      throw new PluginError(`Unable to import ESM module "${modulePath}"`, error)
    }
  }

  if (defaultExport == null) {
    throw new PluginError(
      `Unable to import module "${modulePath}"`,
      new Error(`Couldn't find valid default export`),
    )
  }

  return defaultExport as T
}

/**
 * Searches for a configuration file under the given root directory and loads its content.
 *
 * @param root The root directory to search in.
 * @param baseName The base name of the config file (e.g., `"pages.config"`).
 *
 * @returns An object containing the file path and its parsed data, or `null` if not found.
 */
export const findConfigFile = async (
  root: string,
  baseName: string,
): Promise<{ path: string; data: unknown } | null> => {
  const extensions = ['yaml', 'yml', 'json', 'js', 'cjs', 'mjs']

  for (const ext of extensions) {
    const filePath = path.join(root, `${baseName}.${ext}`)
    if (!fs.existsSync(filePath)) {
      continue
    }

    try {
      switch (ext) {
        case 'yaml':
        case 'yml':
          return {
            path: filePath,
            data: Yaml.load(fs.readFileSync(filePath, 'utf8'), {
              filename: filePath,
              schema: Yaml.FAILSAFE_SCHEMA,
            }),
          }

        case 'json':
          return {
            path: filePath,
            data: JSON.parse(fs.readFileSync(filePath, 'utf8')),
          }

        default:
          return { path: filePath, data: await getDefaultExport(filePath) }
      }
    } catch (error) {
      throw new PluginError(`Failed to load config file "${filePath}"`, error)
    }
  }

  return null
}

/**
 * Wraps a Gatsby API function with error handling.
 *
 * @param api - The name of the Gatsby API being wrapped (e.g., `"onCreatePage"`).
 * @param fn - The original Gatsby API function to be wrapped.
 *
 * @returns The wrapped function with error handling.
 */
export const createPluginExport = <T extends keyof GatsbyNode, F extends GatsbyNode[T]>(
  api: T,
  fn: F,
): F => {
  return ((...args: any[]) => {
    try {
      // Call the original function with provided arguments
      const result = (fn as any)(...args)

      // If the function returns a Promise, attach error handling
      if (result instanceof Promise) {
        return result.catch((error) => reporter.error(api, error))
      }

      // Return the result (for synchronous functions)
      return result
    } catch (error) {
      // Catch synchronous errors and report them
      reporter.error(api, error)
    }
  }) as F
}

/**
 * Custom error class for plugin-related errors.
 */
export class PluginError extends Error {
  /**
   * @param message - A description of the error.
   * @param originalError - The original error that caused this issue (optional).
   */
  constructor(
    message: string,
    public readonly originalError?: Error,
  ) {
    super(message)
  }
}
