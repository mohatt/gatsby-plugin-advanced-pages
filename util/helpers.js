const path = require('path')
const debug = require("debug")("gatsby-plugin-advanced-pages")

// Checks if a path is a child of another one
function isPathChildof (child, parent) {
  const relative = path.relative(parent, child)
  return relative && !relative.startsWith('..') && !path.isAbsolute(relative)
}

module.exports = {
  debug,
  isPathChildof,
}
