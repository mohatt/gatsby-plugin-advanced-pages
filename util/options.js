const path = require('path')
const _ = require("lodash")
const { debug } = require('./helpers')

// Available transformer engines
const ENGINES = {
  mdx: {
    name: 'mdx',
    typeName: 'Mdx',
    bodyFieldName: 'body'
  },
  remark: {
    title: 'remark',
    typeName: 'MarkdownRemark',
    bodyFieldName: 'html'
  }
}

// Initializes options
// The function only runs at first call
// Subsequent calls return the cached options object
const initializeOptions = _.once(({ store, reporter, defaultOptions, userOptions }) => {
  const { program } = store.getState()
  const o = _.merge(defaultOptions, userOptions)

  // Ensure basePath is absolute
  o.basePath = path.join('/', o.basePath)

  // Ensure we have a valid engine
  // Replace the engine name with the engine object from ENGINES map
  o.engine = ENGINES[_.toLower(o.engine)]
  if(!o.engine) {
    throw new TypeError(
      `Invalid engine option provided '${o.engine}' (available engines ` +
      `are '${Object.keys(ENGINES).join(`' or '`)}')`
    )
  }

  // Convert directory paths to absolute ones
  o.directories = _.mapValues(o.directories, dir => (
    path.join(program.directory, dir)
  ))

  // Ensure we have a valid positive number
  o.pagination.limit = parseInt(o.pagination.limit)
  if(o.pagination.limit <= 0 || isNaN(o.pagination.limit)){
    throw new TypeError(`Expected a positive number for 'pagination.limit' option`)
  }

  // Ensure type names are Capitalized
  o.typeNames = _.mapValues(o.typeNames, _.upperFirst)

  // Debug final options object
  debug("Options", o)

  return o
})

// Gets the initialized options object
function getOptions () {
  return initializeOptions()
}

// Gets the value of a single option
function getOption (optionName) {
  return _.get(getOptions(), optionName)
}

module.exports = {
  initializeOptions,
  getOptions,
  getOption
}
