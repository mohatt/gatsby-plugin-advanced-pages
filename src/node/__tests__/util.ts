import path from 'path'
import lodash from 'lodash'
import { mountFile, setupPlugin, resetVFS, projectRoot } from '@test/util'
import { createPluginExport, getDefaultExport, ensurePath, reporter } from '../util'

describe('util', () => {
  beforeEach(() => {
    vi.clearAllMocks()
    vi.resetAllMocks()
    resetVFS()
    setupPlugin()
  })

  describe('createPluginExport', () => {
    const mockApi: any = 'onCreatePage'

    describe.each(['sync', 'async'] as const)('%s function', (type) => {
      it('should execute the original function correctly', async () => {
        const mockFn =
          type === 'sync' ? vi.fn(() => 'return value') : vi.fn(async () => 'async return value')

        const wrappedFn = createPluginExport(mockApi, mockFn)

        if (type === 'sync') {
          let result: any
          expect(() => {
            result = wrappedFn()
          }).not.toThrow()
          expect(result).toBe('return value')
        } else {
          await expect(wrappedFn()).resolves.toBe('async return value')
        }

        expect(mockFn).toHaveBeenCalledTimes(1)
      })

      it('should catch errors and call reporter.error()', async () => {
        const reporterError = new Error('Mocked reporter error')
        vi.spyOn(reporter, 'error').mockImplementation(() => {
          throw reporterError
        })

        const error = new Error('MockFn error')
        const mockFn =
          type === 'sync'
            ? vi.fn(() => {
                throw error
              })
            : vi.fn(async () => {
                throw error
              })

        const wrappedFn = createPluginExport(mockApi, mockFn)

        if (type === 'sync') {
          expect(() => wrappedFn()).toThrow(reporterError)
        } else {
          await expect(wrappedFn()).rejects.toThrow(reporterError)
        }

        expect(mockFn).toHaveBeenCalledTimes(1)
        expect(reporter.error).toHaveBeenCalledWith(mockApi, error)
      })
    })
  })

  describe('getDefaultExport', () => {
    const exportValue = 'DEFAULT_EXPORT'
    const getModulePath = (name: string) => path.join(__dirname, '__fixtures__/files', name)
    const localModules = [
      'test-module.cjs',
      'test-module-default.cjs',
      'test-module.mjs',
      'test-module.json',
    ]

    describe.each(['import', 'require'] as const)('%s', (method) => {
      beforeEach(() => {
        if (method === 'require') {
          vi.stubEnv('VITEST', 'false')
        }
      })

      it.each(localModules)('should correctly load local module (%s)', async (fileName) => {
        await expect(getDefaultExport(getModulePath(fileName))).resolves.toBe(exportValue)
      })

      it('should correctly load external module (lodash)', async () => {
        await expect(getDefaultExport('lodash')).resolves.toBe(lodash)
      })

      it('should correctly load system module (path)', async () => {
        await expect(getDefaultExport('path')).resolves.toBe(path)
      })

      it('throws if the module does not exist', async () => {
        await expect(getDefaultExport('./non-existent.js')).rejects.toMatchSnapshot()
      })

      if (method === 'import') {
        it.each(localModules)('throws for invalid default export (mock: %s)', async (fileName) => {
          const modulePath = getModulePath(fileName)
          vi.doMock(modulePath, () => ({ default: null }))
          await expect(getDefaultExport(modulePath)).rejects.toMatchSnapshot({} as any)
          vi.doUnmock(modulePath)
        })
      }
    })
  })

  describe('ensurePath', () => {
    const cases = [
      {
        case: 'found in parent directory',
        parent: '/mock/parent',
        location: 'file.txt',
        files: ['/mock/parent/file.txt', `${projectRoot}/file.txt`],
      },
      {
        case: 'found in parent directory (absolute)',
        parent: '/mock/parent',
        location: '/to/file.txt',
        files: ['/mock/parent/to/file.txt', `${projectRoot}/to/file.txt`, '/to/file.txt'],
      },
      {
        case: 'found in project root',
        parent: undefined,
        location: 'file.txt',
        files: [`${projectRoot}/file.txt`],
      },
      {
        case: 'found in project root (absolute)',
        parent: undefined,
        location: '/to/file.txt',
        files: [`${projectRoot}/to/file.txt`, '/to/file.txt'],
      },
      {
        case: 'fallback to project root when parent directory is missing',
        parent: '/mock/parent',
        location: 'file.txt',
        files: [`${projectRoot}/file.txt`],
      },
      {
        case: 'found as absolute path',
        parent: '/mock/parent',
        location: '/absolute/file.txt',
        files: ['/absolute/file.txt'],
      },
    ]

    describe.each(cases)('$case', ({ parent, location, files }) => {
      it('should return the correct path if the file exists', () => {
        files.forEach((file) => mountFile(file))
        expect(() => ensurePath(location, parent)).not.toThrow()
        expect(ensurePath(location, parent)).toBe(files[0])
      })
    })

    it.each([['missing.txt'], ['/path/to/missing.txt'], ['/path/to/missing.txt', '/parent']])(
      'throws if the file does not exist (%s)',
      (location, parent) => {
        expect(() => ensurePath(location, parent)).toThrowErrorMatchingSnapshot()
      },
    )
  })
})
