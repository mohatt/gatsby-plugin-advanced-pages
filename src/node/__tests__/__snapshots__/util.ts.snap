// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`util > ensurePath > throws if the file does not exist (/path/to/missing.txt) 1`] = `
[Error: A path with value "/path/to/missing.txt" could not be found at any of the following locations:
 - "/virtual/project/path/to/missing.txt
 - "/path/to/missing.txt"]
`;

exports[`util > ensurePath > throws if the file does not exist (/path/to/missing.txt) 2`] = `
[Error: A path with value "/path/to/missing.txt" could not be found at any of the following locations:
 - "/parent/path/to/missing.txt
 - "/virtual/project/path/to/missing.txt
 - "/path/to/missing.txt"]
`;

exports[`util > ensurePath > throws if the file does not exist (missing.txt) 1`] = `
[Error: A path with value "missing.txt" could not be found at any of the following locations:
 - "/virtual/project/missing.txt"]
`;

exports[`util > getDefaultExport > import > throws for invalid default export (mock: test-module.cjs) 1`] = `
{
  "originalError": [Error: Couldn't find valid default export],
}
`;

exports[`util > getDefaultExport > import > throws for invalid default export (mock: test-module.json) 1`] = `
{
  "originalError": [Error: Couldn't find valid default export],
}
`;

exports[`util > getDefaultExport > import > throws for invalid default export (mock: test-module.mjs) 1`] = `
{
  "originalError": [Error: Couldn't find valid default export],
}
`;

exports[`util > getDefaultExport > import > throws for invalid default export (mock: test-module-default.cjs) 1`] = `
{
  "originalError": [Error: Couldn't find valid default export],
}
`;

exports[`util > getDefaultExport > import > throws if the module does not exist 1`] = `[Error: Unable to import ESM module "./non-existent.js"]`;

exports[`util > getDefaultExport > require > throws if the module does not exist 1`] = `[Error: Unable to import module "./non-existent.js"]`;
